panic: syscall/js: Value.Call: property delete is not a function, got undefined

```
panic: syscall/js: Value.Call: property delete is not a function, got undefined wasm_exec.js:41:14
wasm_exec.js:41:14
goroutine 2 [running]: wasm_exec.js:41:14
syscall/js.Value.Call(0x7ff8000000000801, 0x50d1f, 0x6, 0x46c748, 0x1, 0x1, 0x54f320) wasm_exec.js:41:14
	/home/tp/applications/go/src/syscall/js/js.go:321 +0x86 wasm_exec.js:41:14
github.com/gopherjs/vecty.wrappedObject.Delete(...) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom_wasmjs_gopherjs.go:97 wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).removeProperties(0x5b9c00, 0x55a460) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:305 +0xf wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcileProperties(0x5b9c00, 0x55a460) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:218 +0xbe wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcile(0x5b9c00, 0x55a460, 0x55a460, 0x181c000f, 0x1048000b) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:207 +0x3a wasm_exec.js:41:14
github.com/gopherjs/vecty.render(0x83380, 0x5b9c00, 0x83380, 0x55a460, 0x5b9c00, 0x85201, 0x5b6db8, 0x1, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:981 +0x15 wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcileChildren(0x5b9ce0, 0x55a540, 0x547be8, 0x201801, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:495 +0x7a wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcile(0x5b9ce0, 0x55a540, 0x55a540, 0x51d02, 0xb) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:210 +0x36 wasm_exec.js:41:14
github.com/gopherjs/vecty.render(0x83380, 0x5b9ce0, 0x83380, 0x55a540, 0x0, 0x5b6b01, 0x7ff80003000007fa, 0x5b6b00, 0xaaf5a8) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:981 +0x15 wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcileChildren(0x5beee0, 0x55b420, 0x50bb7, 0x5, 0x85260) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:495 +0x7a wasm_exec.js:41:14
github.com/gopherjs/vecty.KeyedList.reconcile(0x5beee0, 0x0, 0x0, 0x5be2a0, 0x833c0, 0x55e100, 0x5b6d80, 0x1, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:708 +0x44 wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcileChildren(0x5be2a0, 0x55a7e0, 0x4daca0, 0x1, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:467 +0x10d wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcile(0x5be2a0, 0x55a7e0, 0x55a7e0, 0x100000000000001, 0x1048000b) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:210 +0x36 wasm_exec.js:41:14
github.com/gopherjs/vecty.render(0x83380, 0x5be2a0, 0x83380, 0x55a7e0, 0x5be2a0, 0x1, 0x0, 0x0, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:981 +0x15 wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcileChildren(0x5be380, 0x55a8c0, 0x4dabe8, 0x184c0001, 0x7ff80003000007fc) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:495 +0x7a wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcile(0x5be380, 0x55a8c0, 0x55a8c0, 0x1a4, 0x1048000b) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:210 +0x36 wasm_exec.js:41:14
github.com/gopherjs/vecty.render(0x83380, 0x5be380, 0x83380, 0x55a8c0, 0x5be380, 0x5c0201, 0x2, 0x2, 0x85260) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:981 +0x15 wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcileChildren(0x5bee00, 0x55b340, 0x4dab40, 0x5a6f01, 0x82c80) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:495 +0x7a wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcile(0x5bee00, 0x55b340, 0x55b340, 0x0, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:210 +0x36 wasm_exec.js:41:14
github.com/gopherjs/vecty.renderComponent(0xaaff98, 0x495d40, 0x831c0, 0x495d40, 0x5a6ea0, 0x4dab01, 0x5b6ba8, 0x7ff8000000000201, 0x51d02) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:1047 +0x1a wasm_exec.js:41:14
github.com/gopherjs/vecty.render(0x831c0, 0x5a6ea0, 0x831c0, 0x495d40, 0x5b6ba8, 0x1, 0x0, 0x0, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:985 +0x9 wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcileChildren(0x5b16c0, 0x554460, 0x50bb7, 0x5, 0x85260) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:495 +0x7a wasm_exec.js:41:14
github.com/gopherjs/vecty.KeyedList.reconcile(0x5b16c0, 0x0, 0x0, 0x5a90a0, 0x833c0, 0x5510c0, 0x5b61a8, 0x1, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:708 +0x44 wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcileChildren(0x5a90a0, 0x525dc0, 0x41fed8, 0x1, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:467 +0x10d wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcile(0x5a90a0, 0x525dc0, 0x525dc0, 0x5b6178, 0xb) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:210 +0x36 wasm_exec.js:41:14
github.com/gopherjs/vecty.render(0x83380, 0x5a90a0, 0x83380, 0x525dc0, 0x5b6178, 0x1, 0x0, 0x0, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:981 +0x15 wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcileChildren(0x5a9180, 0x525ea0, 0x41fdf8, 0x451701, 0x35ae0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:495 +0x7a wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcile(0x5a9180, 0x525ea0, 0x525ea0, 0x450701, 0x1048000b) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:210 +0x36 wasm_exec.js:41:14
github.com/gopherjs/vecty.render(0x83380, 0x5a9180, 0x83380, 0x525ea0, 0x5a9180, 0x1, 0x0, 0x0, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:981 +0x15 wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcileChildren(0x5a9260, 0x54a000, 0x41f958, 0x5a8d01, 0x82c80) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:495 +0x7a wasm_exec.js:41:14
github.com/gopherjs/vecty.(*HTML).reconcile(0x5a9260, 0x54a000, 0x54a000, 0x0, 0x0) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:210 +0x36 wasm_exec.js:41:14
github.com/gopherjs/vecty.renderComponent(0x84480, 0x40a240, 0xaaf7a0, 0x40a240, 0x40a240, 0x0, 0x109e004b, 0xab56f8, 0x10dd0003) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:1047 +0x1a wasm_exec.js:41:14
github.com/gopherjs/vecty.(*batchRenderer).render(0x200f60, 0x409e9d13879c4113) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:878 +0x22 wasm_exec.js:41:14
github.com/gopherjs/vecty.requestAnimationFrame.func1(0x85260, 0x21a980, 0x5ac1f0, 0x1, 0x1, 0x426e0, 0x21a980) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom.go:1151 +0xb wasm_exec.js:41:14
github.com/gopherjs/vecty.funcOf.func1(0x0, 0x5990f8, 0x1, 0x1, 0x599001, 0x15b0a82e5dc95600) wasm_exec.js:41:14
	/home/tp/projects/gitlab/deepin-code-release-platform/frontend/src/github.com/gopherjs/vecty/dom_wasmjs_gopherjs.go:36 +0x15 wasm_exec.js:41:14
syscall/js.handleEvent() wasm_exec.js:41:14
	/home/tp/applications/go/src/syscall/js/func.go:90 +0x1d wasm_exec.js:41:14
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,8 +1,8 @@
 sudo: false
 install:
-  # Manually download and install Go 1.12 because the Travis / gimme version
-  # is broken (see https://travis-ci.community/t/goos-js-goarch-wasm-go-run-fails-panic-newosproc-not-implemented/1651/6)
-  - wget -O go.tar.gz https://dl.google.com/go/go1.12.linux-amd64.tar.gz
+  # Manually download and install Go 1.14 because the Travis / gimme version
+  # has been broken in the past (see https://travis-ci.community/t/goos-js-goarch-wasm-go-run-fails-panic-newosproc-not-implemented/1651/6)
+  - wget -O go.tar.gz https://dl.google.com/go/go1.14.linux-amd64.tar.gz
   - tar -C ~ -xzf go.tar.gz
   - rm go.tar.gz
   - export GOROOT=~/go
@@ -57,7 +57,8 @@ script:
   # Test with Go compiler (under amd64 and wasm architectures) and GopherJS compiler.
   - go test -race ./...
   - GO111MODULE=on GOOS=js GOARCH=wasm go test -exec="$(go env GOROOT)/misc/wasm/go_js_wasm_exec" ./...
-  - gopherjs test ./...
+  # Disabled until https://github.com/gopherjs/gopherjs/issues/962 is fixed.
+  #- gopherjs test ./...
 
   # Generate and upload coverage to codecov.io
   - goverage -covermode=atomic -coverprofile=coverage.out $(go list ./... | grep -v -e vecty/elem -e vecty/event -e vecty/example -e vecty/prop -e vecty/style)

--- a/.travis.yml
+++ b/.travis.yml
@@ -21,7 +21,7 @@ install:
   - npm install --global node-gyp
 
   # Linters, etc.
-  - curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s -- -b $(go env GOPATH)/bin v1.17.1
+  - GO111MODULE=off go get -u github.com/golangci/golangci-lint/cmd/golangci-lint
   - go get -u github.com/haya14busa/goverage
   - go get -u mvdan.cc/gofumpt
   - go get -u mvdan.cc/gofumpt/gofumports

--- a/doc/CHANGELOG.md
+++ b/doc/CHANGELOG.md
@@ -6,6 +6,14 @@ Although v1.0.0 [is not yet out](https://github.com/gopherjs/vecty/milestone/1),
 Pre-v1.0.0 Breaking Changes
 ---------------------------
 
+## February 28, 2020 ([PR #256](https://github.com/gopherjs/vecty/pull/256)): indirect breaking change
+
+- Go 1.14+ is now required by Vecty. Users of older Go versions and/or GopherJS (until https://github.com/gopherjs/gopherjs/issues/962 is fixed) may wish to continue using commit `6a0a25ee5a96ce029e684c7da6333aa1f34f8f96`.
+
+## Nov 30, 2019 ([PR #249](https://github.com/gopherjs/vecty/pull/249)): minor breaking change
+
+- `vecty.RenderBody(comp)` is now a blocking function call. Users that rely on it being non-blocking can instead now use `if err := vecty.RenderInto("body", comp); err != nil { panic(err) }`
+
 ## June 30, 2019 ([PR #232](https://github.com/gopherjs/vecty/pull/232)): major breaking change
 
 - `(*HTML).Node` now returns a `syscall/js.Value` instead of `*gopherjs/js.Object`. Users will need to update to the new `syscall/js` API in their applications.

--- a/dom.go
+++ b/dom.go
@@ -202,7 +202,7 @@ func (h *HTML) reconcile(prev *HTML) []Mounter {
 		h.createNode()
 	}
 
-	if h.node != prev.node {
+	if !h.node.Equal(prev.node) {
 		// reconcile properties against empty prev for new nodes.
 		h.reconcileProperties(&HTML{})
 	} else {
@@ -216,7 +216,7 @@ func (h *HTML) reconcile(prev *HTML) []Mounter {
 // element.
 func (h *HTML) reconcileProperties(prev *HTML) {
 	// If nodes match, remove any outdated properties
-	if h.node == prev.node {
+	if h.node.Equal(prev.node) {
 		h.removeProperties(prev)
 	}
 
@@ -374,7 +374,7 @@ func (h *HTML) reconcileChildren(prev *HTML) (pendingMounts []Mounter) {
 		//
 		// TODO(pdf): Add tests for node equality, keyed children
 		var (
-			new     = h.node != prev.node
+			new     = !h.node.Equal(prev.node)
 			nextKey interface{}
 		)
 		keyer, isKeyer := nextChild.(Keyer)
@@ -459,7 +459,7 @@ func (h *HTML) reconcileChildren(prev *HTML) (pendingMounts []Mounter) {
 		}
 		// If our insertion node is the current previous child, advance to the
 		// next sibling.
-		if prevChildRender != nil && prevChildRender.node == h.insertBeforeNode {
+		if prevChildRender != nil && prevChildRender.node.Equal(h.insertBeforeNode) {
 			h.insertBeforeNode = h.insertBeforeNode.Get("nextSibling")
 		}
 
@@ -487,7 +487,7 @@ func (h *HTML) reconcileChildren(prev *HTML) (pendingMounts []Mounter) {
 			insertBeforeKeyedNode = h.lastRenderedChild.nextSibling()
 			// If the next node is our old node, mark key as stable, to avoid
 			// unnecessary insertion.
-			if prevChildRender != nil && prevChildRender.node == insertBeforeKeyedNode {
+			if prevChildRender != nil && prevChildRender.node.Equal(insertBeforeKeyedNode) {
 				stableKey = true
 				insertBeforeKeyedNode = nil
 			}
@@ -533,7 +533,7 @@ func (h *HTML) reconcileChildren(prev *HTML) (pendingMounts []Mounter) {
 				pendingMounts = append(pendingMounts, m)
 			}
 
-			if hasKeyedChildren && (prevChildRender != nil && prevChildRender.node == nextChildRender.node) {
+			if hasKeyedChildren && (prevChildRender != nil && prevChildRender.node.Equal(nextChildRender.node)) {
 				// We are re-using the name node. Remove the children from
 				// keyedChildren so that we don't remove it when we remove dangling
 				// children below.
@@ -631,7 +631,7 @@ func (h *HTML) nextSibling() jsObject {
 func (h *HTML) removeChild(child *HTML) {
 	// If we're removing the current insert target, use the next
 	// sibling, if any.
-	if h.insertBeforeNode != nil && h.insertBeforeNode == child.node {
+	if h.insertBeforeNode != nil && h.insertBeforeNode.Equal(child.node) {
 		h.insertBeforeNode = h.insertBeforeNode.Get("nextSibling")
 	}
 	unmount(child)
@@ -1080,7 +1080,7 @@ func mountUnmount(next, prev ComponentOrHTML) Mounter {
 		return nil
 	}
 	if prevHTML := extractHTML(prev); prevHTML != nil {
-		if nextHTML := extractHTML(next); nextHTML == nil || prevHTML.node != nextHTML.node {
+		if nextHTML := extractHTML(next); nextHTML == nil || !prevHTML.node.Equal(nextHTML.node) {
 			for _, child := range prevHTML.children {
 				unmount(child)
 			}
@@ -1279,6 +1279,8 @@ type jsObject interface {
 	Call(name string, args ...interface{}) jsObject
 	String() string
 	Truthy() bool
+	Equal(other jsObject) bool
+	IsUndefined() bool
 	Bool() bool
 	Int() int
 	Float() float64
@@ -1293,7 +1295,7 @@ func init() {
 	if global == nil {
 		panic("vecty: only GopherJS and WebAssembly compilation is supported")
 	}
-	if global.Get("document") == undefined {
+	if global.Get("document").IsUndefined() {
 		panic("vecty: only running inside a browser is supported")
 	}
 }

--- a/dom_native.go
+++ b/dom_native.go
@@ -45,7 +45,7 @@ func toLower(s string) string {
 
 var (
 	global    jsObject
-	undefined wrappedObject
+	undefined = wrappedObject{j: &jsObjectImpl{}}
 )
 
 func funcOf(fn func(this jsObject, args []jsObject) interface{}) jsFunc {
@@ -64,6 +64,14 @@ func valueOf(v interface{}) jsObject { return valueOfImpl(v) }
 type wrappedObject struct {
 	jsObject
 	j jsObject
+}
+
+type jsObjectImpl struct {
+	jsObject
+}
+
+func (e *jsObjectImpl) Equal(other jsObject) bool {
+	return e == other.(*jsObjectImpl)
 }
 
 var (

--- a/dom_test.go
+++ b/dom_test.go
@@ -79,7 +79,7 @@ func TestHTML_Node(t *testing.T) {
 
 	x := undefined
 	h := &HTML{node: x}
-	if h.Node() != x.j {
+	if !h.Node().Equal(x.j) {
 		t.Fatal("h.Node() != x")
 	}
 }

--- a/dom_wasmjs_gopherjs.go
+++ b/dom_wasmjs_gopherjs.go
@@ -38,8 +38,8 @@ func toLower(s string) string {
 }
 
 var (
-	global    = wrapObject(js.Global())
-	undefined = wrappedObject{js.Undefined()}
+	global = wrapObject(js.Global())
+	undefined = wrapObject(js.Undefined())
 )
 
 func funcOf(fn func(this jsObject, args []jsObject) interface{}) jsFunc {
@@ -73,11 +73,8 @@ func valueOf(v interface{}) jsObject {
 }
 
 func wrapObject(j js.Value) jsObject {
-	if j == js.Null() {
+	if j.IsNull() {
 		return nil
-	}
-	if j == js.Undefined() {
-		return undefined
 	}
 	return wrappedObject{j: j}
 }
@@ -105,7 +102,7 @@ func (w wrappedObject) Get(key string) jsObject {
 }
 
 func (w wrappedObject) Delete(key string) {
-	w.j.Call("delete", key)
+	w.j.Delete(key)
 }
 
 func (w wrappedObject) Call(name string, args ...interface{}) jsObject {
@@ -121,6 +118,17 @@ func (w wrappedObject) String() string {
 
 func (w wrappedObject) Truthy() bool {
 	return w.j.Truthy()
+}
+
+func (w wrappedObject) IsUndefined() bool {
+	return w.j.IsUndefined()
+}
+
+func (w wrappedObject) Equal(other jsObject) bool {
+	if w.j.IsNull() != (other == nil) {
+		return false
+	}
+	return w.j.Equal(unwrap(other).(js.Value))
 }
 
 func (w wrappedObject) Bool() bool {

--- a/dom_wasmjs_gopherjs.go
+++ b/dom_wasmjs_gopherjs.go
@@ -38,7 +38,7 @@ func toLower(s string) string {
 }
 
 var (
-	global = wrapObject(js.Global())
+	global    = wrapObject(js.Global())
 	undefined = wrapObject(js.Undefined())
 )
 

--- a/dom_wasmjs_gopherjs.go
+++ b/dom_wasmjs_gopherjs.go
@@ -39,7 +39,7 @@ func toLower(s string) string {
 
 var (
 	global    = wrapObject(js.Global())
-	undefined = wrapObject(js.Undefined())
+	undefined = wrappedObject{js.Undefined()}
 )
 
 func funcOf(fn func(this jsObject, args []jsObject) interface{}) jsFunc {

--- a/domutil.go
+++ b/domutil.go
@@ -1,7 +1,7 @@
 package vecty
 
 func replaceNode(newNode, oldNode jsObject) {
-	if newNode == oldNode {
+	if newNode.Equal(oldNode) {
 		return
 	}
 	oldNode.Get("parentNode").Call("replaceChild", newNode, oldNode)

--- a/example/todomvc/example.go
+++ b/example/todomvc/example.go
@@ -35,7 +35,7 @@ func attachLocalStorage() {
 		js.Global().Get("localStorage").Set("items", string(data))
 	})
 
-	if data := js.Global().Get("localStorage").Get("items"); data != js.Undefined() {
+	if data := js.Global().Get("localStorage").Get("items"); !data.IsUndefined() {
 		var items []*model.Item
 		if err := json.Unmarshal([]byte(data.String()), &items); err != nil {
 			println("failed to load items: " + err.Error())

--- a/testsuite_test.go
+++ b/testsuite_test.go
@@ -96,7 +96,7 @@ func (v *valueMocker) get(invocation string) interface{} {
 type testSuiteT struct {
 	t                                      *testing.T
 	callbacks                              map[string]interface{}
-	strings, bools, floats, ints, truthies *valueMocker
+	strings, bools, floats, ints, truthies, isUndefined *valueMocker
 
 	got    string
 	isDone bool
@@ -274,6 +274,9 @@ func (r *objectRecorder) IsUndefined() bool { return r.ts.isUndefined.get(r.name
 
 // Equal implements the jsObject interface.
 func (r *objectRecorder) Equal(other jsObject) bool {
+	if (r == nil) != (other == nil) {
+		return false
+	}
 	return r == other.(*objectRecorder)
 }
 

--- a/testsuite_test.go
+++ b/testsuite_test.go
@@ -47,13 +47,13 @@ func TestMain(m *testing.M) {
 
 func testSuite(t *testing.T) *testSuiteT {
 	ts := &testSuiteT{
-		t:         t,
-		callbacks: make(map[string]interface{}),
-		strings:   &valueMocker{},
-		bools:     &valueMocker{},
-		floats:    &valueMocker{},
-		ints:      &valueMocker{},
-		truthies:  &valueMocker{},
+		t:           t,
+		callbacks:   make(map[string]interface{}),
+		strings:     &valueMocker{},
+		bools:       &valueMocker{},
+		floats:      &valueMocker{},
+		ints:        &valueMocker{},
+		truthies:    &valueMocker{},
 		isUndefined: &valueMocker{},
 	}
 	global = &objectRecorder{

--- a/testsuite_test.go
+++ b/testsuite_test.go
@@ -94,8 +94,8 @@ func (v *valueMocker) get(invocation string) interface{} {
 }
 
 type testSuiteT struct {
-	t                                      *testing.T
-	callbacks                              map[string]interface{}
+	t                                                   *testing.T
+	callbacks                                           map[string]interface{}
 	strings, bools, floats, ints, truthies, isUndefined *valueMocker
 
 	got    string

--- a/testsuite_test.go
+++ b/testsuite_test.go
@@ -54,6 +54,7 @@ func testSuite(t *testing.T) *testSuiteT {
 		floats:    &valueMocker{},
 		ints:      &valueMocker{},
 		truthies:  &valueMocker{},
+		isUndefined: &valueMocker{},
 	}
 	global = &objectRecorder{
 		ts:   ts,
@@ -267,6 +268,14 @@ func (r *objectRecorder) String() string { return r.ts.strings.get(r.name).(stri
 
 // Truthy implements the jsObject interface.
 func (r *objectRecorder) Truthy() bool { return r.ts.truthies.get(r.name).(bool) }
+
+// IsUndefined implements the jsObject interface.
+func (r *objectRecorder) IsUndefined() bool { return r.ts.isUndefined.get(r.name).(bool) }
+
+// Equal implements the jsObject interface.
+func (r *objectRecorder) Equal(other jsObject) bool {
+	return r == other.(*objectRecorder)
+}
 
 // Bool implements the jsObject interface.
 func (r *objectRecorder) Bool() bool { return r.ts.bools.get(r.name).(bool) }

